[1.0.0-alpha2] Workload and component (#316)
oam-dev/spec: core/workloads/containerized_workload/containerized_workload.md

# Containerized Workload

> A containerized workload is used for long-running workloads in containers. Here is the definition of the containerized workload.

Reviewer: This is where execution policies are needed. A containerized workload need not strictly be a long-running workload.

Author: IMO, there are two ways to solve this.

- One is to rename this to be `containerized server` and add another workload called `containerized task` which has a different schema (task specific) from server. 

- The other is to keep this as the only core containerized workload and add policies for components.

These two ways can actually co-exist too.

Reviewer 2: > which has a different schema (task specific) from server

How do we expect task specific schema would differ? Under the contemporary spec the schema of task and server (and worker) are identical.

Reviewer: I don't think they would differ at all. That's the problem we solved with workloadType in OAM alpha1 and have now re-introduced here. Rather than define multiple workload definitions with identical schemas, I think we need a way to classify execution behavior or something like that, so that we avoid duplicate schema definitions and have a way to express whether or not a trait is applicable to a given component.

Reviewer 2: @vturecek To be honest it seems pretty reasonable to me for workloads that are conceptually different things to encode that difference in their GVK, even if they have identical schema. Previously we effectively had a `ComponentSchematic` type with a subtype encoded in `spec.workloadType`. Now we have a `Component` type with a subtype encoded in `spec.workload.kind`.

I could see a case for moving the encoding of whether or not something was a singleton into the schema as a field under `spec` and simply having three kinds with identical schemas; probably `ContainerizedServer`, `ContainerizedWorker`, and `ContainerizedTask` similar to what @ryanzhang-oss suggested.

Author: > simply having three kinds with identical schemas; probably `ContainerizedServer`, `ContainerizedWorker`, and `ContainerizedTask` similar to what @ryanzhang-oss suggested.

Since those are all core workloads, we need to spell out their schematic in the spec. It seems we either 
- have 3 identical schemas in the spec 
- explain that we use the schema's name to differentiate the runtime characteristics of a workload from another workload with an identical schema.

I think the second approach is pretty much the same as in spec alphav1 as @negz says. I am just not sure how can we evolve it to the policy approach which I think we all agree is the right direction.

Let's find a time to discuss this in a meeting.

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: WorkloadDefinition
metadata:

>   name: containerizedworkload.core.oam.dev

Reviewer 2: ```suggestion
  name: containerizedworkloads.core.oam.dev
```

Author: why using purls? is it K8s convention? We want to keep the two "names" the same.  Does this mean that the CRD name needs purls too?

Reviewer 2: https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#create-a-customresourcedefinition

> name must match the spec fields below, and be in the form: <plural>.<group>

I agree that we want to keep the two names the same, and we're subject to a Kubernetes constraint here. Kubernetes requires that the CRD defining `kind: ContainerizedWorkload` in `group: core.oam.dev` be named `containerizedworkloads.core.oam.dev` (plural). So if we want to refer to a CRD by name (and thus make getting the CRD way easier for our controllers) we must use the plural name.

spec:
  definitionRef:

>     schemaName: containerizedworkload.core.oam.dev

Reviewer 2: ```suggestion
    name: containerizedworkloads.core.oam.dev
```

>     schemaName: containerizedworkload.core.oam.dev

Reviewer 3: ```suggestion
    name: containerizedworkload.core.oam.dev
```
per: https://github.com/crossplane/crossplane/pull/1276#discussion_r383647996

Reviewer 3: See this was already pointed out while I was reviewing, resolving in favor of https://github.com/oam-dev/spec/pull/316#discussion_r384720593

```

> Here is the schematics of the containerized workload. 

Reviewer 2: ```suggestion
Here is the schematic of a containerized workload. 
```

### Top-Level Attributes of a containerized workload 

These attributes provide top-level information about the containerized workload.

| Attribute | Type | Required | Default Value | Description |
|-----------|------|----------|---------------|-------------|
| `apiVersion` | `string` | Y ||  `core.oam.dev/v1alpha2` |

> | `kind` | `string` | Y ||  `containerWorkload` |

Reviewer 2: ```suggestion
| `kind` | `string` | Y ||  `ContainerizedWorkload` |
```
